File: src/main/java/com/github/cao/awa/kalmia/reflection/field/FieldGet.java
package com.github.cao.awa.kalmia.reflection.field;

import com.github.cao.awa.apricot.util.collection.ApricotCollectionFactor;
import com.github.cao.awa.sinuatum.manipulate.Manipulate;

import java.util.List;
import java.util.function.Function;

public class FieldGet<R, I> {
    private final Class<I> target;
    private final List<String> gets = ApricotCollectionFactor.arrayList();
    private final List<Function<Class<I>, R>> backups = ApricotCollectionFactor.arrayList();

    private FieldGet(I target) {
        this.target = Manipulate.cast(target.getClass());
    }

    private FieldGet(Class<I> target) {
        this.target = Manipulate.cast(target);
    }

    public static <X, Y> FieldGet<X, Y> create(Y target, String name) {
        return new FieldGet<X, Y>(target).or(name);
    }

    public static <X, Y> FieldGet<X, Y> create(Class<Y> target, String name) {
        return new FieldGet<X, Y>(target).or(name);
    }

    public FieldGet<R, I> or(String name) {
        this.gets.add(name);
        return this;
    }

    public FieldGet<R, I> or(Function<Class<I>, R> function) {
        this.backups.add(function);
        return this;
    }

    @SuppressWarnings("unchecked")
    public <Z> Z get() {
        Class<I> clazz = Manipulate.cast(this.target);

        assert clazz != null;
        for (String name : this.gets) {
            try {
                Object o = clazz.getField(name)
                                .get(this.target);

                if (o != null) {
                    return (Z) o;
                }
            } catch (Exception e) {

            }
        }

        for (Function<Class<I>, R> supplier : this.backups) {
            try {
                R r = supplier.apply(clazz);

                if (r != null) {
                    return (Z) r;
                }
            } catch (Exception e) {

            }
        }

        return null;
    }
}

File: src/main/java/com/github/cao/awa/kalmia/reflection/field/FieldGet.kt
package com.github.cao.awa.kalmia.reflection.field

import com.github.cao.awa.apricot.util.collection.ApricotCollectionFactor
import com.github.zhuaidadaya.rikaishinikui.handler.universal.entrust.EntrustEnvironment

import java.util.function.Function

class FieldGet<R, I> {
    private var target: Class<I>
    private val gets: MutableList<String> = ApricotCollectionFactor.arrayList()
    private val backups: MutableList<Function<Class<I>, R>> = ApricotCollectionFactor.arrayList()

    private constructor(target: I) {
        this.target =
            EntrustEnvironment.cast((target ?: throw RuntimeException())::class.java) ?: throw RuntimeException()
    }

    private constructor(target: Class<I>) {
        this.target = EntrustEnvironment.cast(target) ?: throw RuntimeException()
    }

    fun or(name: String): FieldGet<R, I> {
        this.gets.add(name)
        return this
    }

    fun or(function: Function<Class<I>, R>): FieldGet<R, I> {
        this.backups.add(function)
        return this
    }

    fun <Z> get(): Z? {
        val clazz: Class<I> = EntrustEnvironment.cast<Class<I>>(this.target) ?: throw RuntimeException()

        for (name in this.gets) {
            try {
                val o: Any? = clazz.getField(name).get(this.target)

                if (o != null) {
                    // TODO
                    return o as? Z
                }
            } catch (_: Exception) {

            }
        }

        for (supplier in backups) {
            try {
                val r: R = supplier.apply(clazz)

                if (r != null) {
                    // TODO
                    return r as Z
                }
            } catch (_: Exception) {

            }
        }

        return null
    }

    companion object {
        @JvmStatic
        fun <X, Y> create(target: Y, name: String): FieldGet<X, Y> {
            return FieldGet<X, Y>(target).or(name)
        }

        @JvmStatic
        fun <X, Y> create(target: Class<Y>, name: String): FieldGet<X, Y> {
            return FieldGet<X, Y>(target).or(name)
        }
    }
}
